File: js/test/algorithm/CentroidTest.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { join } from "node:path";
import { getCentroid } from "../../src/algorithm/Centroid.ts";
import type { Coordinate } from "../../src/GeometryAdapter.ts";
import { parseXmlTestCases } from "../utils/XmlTestParser.ts";

const FIXTURE = join(import.meta.dirname, "../../../upstream/jts/resources/testxml/general/TestCentroid.xml");

describe("Centroid", () => {
  const cases = parseXmlTestCases(FIXTURE, "getCentroid");

  it("loads all 38 upstream cases", () => {
    assert.equal(cases.length, 38);
  });

  for (const c of cases) {
    it(`matches JTS for: ${c.desc}`, () => {
      const actual = c.input === null ? null : getCentroid(c.input);
      // Exact comparison, per the exact-comparison rule: both languages evaluate the same IEEE
      // 754 operations in the same order, so any difference is information.
      assert.deepEqual(actual, c.expected);
    });
  }

  it("treats zero-length lines as points", () => {
    // JTS 1.19.0 gives (6.666666666666667, 6.666666666666667) here; the
    // pre-retrofit inline centroid gave (0, 0).
    const centroid = getCentroid({
      type: "MultiLineString",
      coordinates: [
        [
          [0, 0],
          [0, 0],
        ],
        [
          [10, 10],
          [10, 10],
        ],
        [
          [10, 10],
          [10, 10],
        ],
      ],
    });
    assert.notEqual(centroid, null);
    // Both ordinates land on 20/3. The band is the one this case has always been
    // held to; it is tighter than TOLERANCE below, which tracks JTS's own constant.
    const BAND = 5e-13;
    assert.ok(Math.abs(centroid![0] - 6.666666666666667) < BAND);
    assert.ok(Math.abs(centroid![1] - 6.666666666666667) < BAND);
  });

  it("returns null for an empty geometry", () => {
    assert.equal(getCentroid({ type: "MultiPoint", coordinates: [] }), null);
  });

  /** @jts-adapter CentroidTest#TOLERANCE */
  const TOLERANCE = 1e-10;

  /**
   * The area of a ring, transcribed from JTS `Area.ofRing(Coordinate[])` — which
   * is what `Geometry.getArea()` calls. Test-local: no ported source module
   * needs `Geometry.getArea()`, so it does not belong in the adapter.
   *
   * The translation by `x0` is load-bearing, not a micro-optimisation. This
   * test's rings are slivers whose coordinates differ only around the 12th
   * decimal place, so the textbook shoelace form
   * `x[i] * y[i + 1] - x[i + 1] * y[i]` loses every significant digit to
   * cancellation: it returns exactly 0 for two of the three rings here and
   * overstates the third by eleven orders of magnitude.
   *
   * @jts-adapter Geometry.getArea()
   */
  function ringArea(ring: Coordinate[]): number {
    if (ring.length < 3) return 0.0;
    let sum = 0.0;
    const x0 = ring[0][0];
    for (let i = 1; i < ring.length - 1; i++) {
      const x = ring[i][0] - x0;
      const y1 = ring[i + 1][1];
      const y2 = ring[i - 1][1];
      sum += x * (y2 - y1);
    }
    return Math.abs(sum / 2.0);
  }

  /** @jts CentroidTest#areaWeightedCentroid(Geometry) */
  function areaWeightedCentroid(polys: Coordinate[][][]): Coordinate {
    const totalArea = polys.reduce((acc, rings) => acc + ringArea(rings[0]), 0);
    let cx = 0;
    let cy = 0;
    for (const rings of polys) {
      const areaFraction = ringArea(rings[0]) / totalArea;
      const componentCentroid = getCentroid({ type: "Polygon", coordinates: rings })!;
      cx += areaFraction * componentCentroid[0];
      cy += areaFraction * componentCentroid[1];
    }
    return [cx, cy];
  }

  /** @jts CentroidTest#testCentroidMultiPolygon() */
  it("computes a MultiPolygon centroid as the area-weighted average of its components", () => {
    // Verify that the computed centroid of a MultiPolygon is equivalent to the
    // area-weighted average of its components.
    const polys: Coordinate[][][] = [
      [
        [
          [-92.661322, 36.58994900000003],
          [-92.66132199999993, 36.58994900000005],
          [-92.66132199999993, 36.589949000000004],
          [-92.661322, 36.589949],
          [-92.661322, 36.58994900000003],
        ],
      ],
      [
        [
          [-92.65560500000008, 36.58708800000005],
          [-92.65560499999992, 36.58708800000005],
          [-92.65560499998745, 36.587087999992576],
          [-92.655605, 36.587088],
          [-92.65560500000008, 36.58708800000005],
        ],
      ],
      [
        [
          [-92.65512450000065, 36.586800000000466],
          [-92.65512449999994, 36.58680000000004],
          [-92.65512449998666, 36.5867999999905],
          [-92.65512450000065, 36.586800000000466],
        ],
      ],
    ];
    const expected = areaWeightedCentroid(polys);
    const actual = getCentroid({ type: "MultiPolygon", coordinates: polys })!;
    assert.ok(Math.abs(actual[0] - expected[0]) < TOLERANCE);
    assert.ok(Math.abs(actual[1] - expected[1]) < TOLERANCE);
  });
});

File: rs/core/src/test/algorithm/centroid_test.rs
//! Ports `CentroidTest.java`.
//!
//! @jts-deviate CentroidTest — `Centroid` is crate-internal, so `rs/core/tests/`
//!   (an external crate) cannot reach it. The `TestCentroid.xml`-driven test lives
//!   under `src/test/algorithm/` instead, in its own file: an integration test
//!   links against the library built without `cfg(test)`, so it cannot see the
//!   `#[cfg(test)]` locator stack this test needs.

use geo_types::{Coord, Geometry, LineString, MultiLineString, MultiPoint, MultiPolygon, Polygon};

use crate::algorithm::centroid::get_centroid;

// The parser lives in the integration-test crate, which a `#[cfg(test)]`
// module inside `src/` cannot `use`. `#[path] mod` cannot reach it either:
// its base directory would be a directory that does not exist. `include!`
// resolves against this file's own directory, `core/src/test/algorithm/`,
// so it works.
mod xml_test_parser {
    include!("../../../tests/utils/xml_test_parser.rs");
}
use xml_test_parser::parse_xml_test_cases;

const FIXTURE: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../upstream/jts/resources/testxml/general/TestCentroid.xml"
);

#[test]
fn matches_jts_on_every_upstream_case() {
    let cases = parse_xml_test_cases(FIXTURE, "getCentroid");
    assert_eq!(cases.len(), 38, "TestCentroid.xml should yield 38 cases");
    for case in &cases {
        let actual = case.input.as_ref().and_then(get_centroid);
        // Exact comparison, per the exact-comparison rule.
        assert_eq!(actual, case.expected, "case: {}", case.desc);
    }
}

#[test]
fn treats_zero_length_lines_as_points() {
    // Zero-length-line centroid defect: JTS 1.19.0 gives (6.666…, 6.666…) here; the
    // pre-retrofit inline centroid gave (0, 0).
    let mls = MultiLineString(vec![
        LineString::from(vec![(0.0, 0.0), (0.0, 0.0)]),
        LineString::from(vec![(10.0, 10.0), (10.0, 10.0)]),
        LineString::from(vec![(10.0, 10.0), (10.0, 10.0)]),
    ]);
    let c = get_centroid(&Geometry::MultiLineString(mls)).unwrap();
    assert!((c.x - 6.666_666_666_666_667).abs() < 1e-12);
    assert!((c.y - 6.666_666_666_666_667).abs() < 1e-12);
}

#[test]
fn returns_none_for_an_empty_geometry() {
    assert_eq!(
        get_centroid(&Geometry::MultiPoint(MultiPoint(vec![]))),
        None
    );
}

/// @jts-adapter CentroidTest#TOLERANCE
const TOLERANCE: f64 = 1e-10;

/// The area of a ring, transcribed from JTS `Area.ofRing(Coordinate[])` —
/// which is what `Geometry.getArea()` calls. Test-local: no ported source
/// module needs `Geometry.getArea()`, so it does not belong in the adapter.
///
/// The translation by `x0` is load-bearing, not a micro-optimisation. This
/// test's rings are slivers whose coordinates differ only around the 12th
/// decimal place, so the textbook shoelace form
/// `x[i] * y[i + 1] - x[i + 1] * y[i]` loses every significant digit to
/// cancellation: it returns exactly 0 for two of the three rings here and
/// overstates the third by eleven orders of magnitude.
///
/// @jts-adapter Geometry.getArea()
fn ring_area(ring: &[Coord<f64>]) -> f64 {
    if ring.len() < 3 {
        return 0.0;
    }
    let mut sum = 0.0;
    let x0 = ring[0].x;
    for i in 1..ring.len() - 1 {
        let x = ring[i].x - x0;
        let y1 = ring[i + 1].y;
        let y2 = ring[i - 1].y;
        sum += x * (y2 - y1);
    }
    (sum / 2.0).abs()
}

/// @jts CentroidTest#areaWeightedCentroid(Geometry)
fn area_weighted_centroid(polys: &[Polygon<f64>]) -> Coord<f64> {
    let total_area: f64 = polys.iter().map(|p| ring_area(&p.exterior().0)).sum();
    let mut cx = 0.0;
    let mut cy = 0.0;
    for poly in polys {
        let area_fraction = ring_area(&poly.exterior().0) / total_area;
        let component_centroid = get_centroid(&Geometry::Polygon(poly.clone())).unwrap();
        cx += area_fraction * component_centroid.x;
        cy += area_fraction * component_centroid.y;
    }
    Coord { x: cx, y: cy }
}

/// @jts CentroidTest#testCentroidMultiPolygon()
#[test]
fn computes_a_multipolygon_centroid_as_the_area_weighted_average() {
    // Verify that the computed centroid of a MultiPolygon is equivalent to
    // the area-weighted average of its components.
    let polys = vec![
        Polygon::new(
            LineString::from(vec![
                (-92.661322, 36.589_949_000_000_03),
                (-92.661_321_999_999_93, 36.589_949_000_000_05),
                (-92.661_321_999_999_93, 36.589_949_000_000_004),
                (-92.661322, 36.589949),
                (-92.661322, 36.589_949_000_000_03),
            ]),
            vec![],
        ),
        Polygon::new(
            LineString::from(vec![
                (-92.655_605_000_000_08, 36.587_088_000_000_05),
                (-92.655_604_999_999_92, 36.587_088_000_000_05),
                (-92.655_604_999_987_45, 36.587_087_999_992_576),
                (-92.655605, 36.587088),
                (-92.655_605_000_000_08, 36.587_088_000_000_05),
            ]),
            vec![],
        ),
        Polygon::new(
            LineString::from(vec![
                (-92.655_124_500_000_65, 36.586_800_000_000_466),
                (-92.655_124_499_999_94, 36.586_800_000_000_04),
                (-92.655_124_499_986_66, 36.586_799_999_990_5),
                (-92.655_124_500_000_65, 36.586_800_000_000_466),
            ]),
            vec![],
        ),
    ];
    let expected = area_weighted_centroid(&polys);
    let actual = get_centroid(&Geometry::MultiPolygon(MultiPolygon(polys.clone()))).unwrap();
    assert!((actual.x - expected.x).abs() < TOLERANCE);
    assert!((actual.y - expected.y).abs() < TOLERANCE);
}
